msprime/ancestry.py

```diff
@@ -721,113 +762,375 @@ def simulate(
+            sequence_length = initial_state.sequence_length
+        elif recombination_rate is None and gene_conversion_rate is None:
+            # In this case, we're doing single-locus simulations, so a sequence
+            # length of 1 makes sense.
             sequence_length = 1
+        elif isinstance(recombination_rate, intervals.RateMap):
+            sequence_length = recombination_rate.sequence_length
+        elif isinstance(gene_conversion_rate, intervals.RateMap):
+            sequence_length = gene_conversion_rate.sequence_length
+        else:
+            raise ValueError(
+                "A sequence_length value must be specified. This can be either "
+                "via the the sequence_length parameter itself, of implicitly "
+                "through using a RateMap instance for the recombination_rate "
+                "or gene_conversion_rate parameters, or via the initial_state "
+                "tables. "
+            )
     else:
-        # The recombination_rate argument is more flexible.
-        if isinstance(recombination_rate, intervals.RateMap):
-            recomb_map = recombination_rate
-            recombination_rate = None
+        sequence_length = float(sequence_length)
+    assert sequence_length is not None
+
+    if discrete_genome and math.floor(sequence_length) != sequence_length:
+        raise ValueError("Must have integer sequence length with discrete_genome=True")
+
+    recombination_map = _parse_rate_map(
+        recombination_rate, sequence_length, "recombination"
+    )
+    gene_conversion_map = _parse_rate_map(
+        gene_conversion_rate, sequence_length, "gene conversion"
+    )
+    if gene_conversion_track_length is None:
+        if gene_conversion_rate is None:
+            # It doesn't matter what the track_length is, just set a
+            # value to keep the low-level code happy.
+            gene_conversion_track_length = 1
         else:
-            if sequence_length is None:
+            raise ValueError(
+                "Must specify track length when simulating gene conversion"
+            )
+    else:
+        if gene_conversion_rate is None:
+            raise ValueError(
+                "Must specify gene conversion rate along with track length"
+            )
+        gene_conversion_track_length = float(gene_conversion_track_length)
+
+    # TODO we many need to have different ploidy values for the different models.
+    # For example, dtwf is an inherently diploid process and anything else doesn't
+    # make sense. However, having different defaults for ploidy will lead to
+    # confusing problems when we have mixed models, so perhaps we should raise
+    # an error from models that require ploidy=2 and force them to explicitly
+    # set it.
+    # TODO Might also be worth considering setting ploidy=2 by default. This
+    # *is* likely to be the most common application, and those doing haploid
+    # stuff can set ploidy=1, if they like.
```

Comment:
I support a default of ploidy=2.

Reply:
I agree. It also makes the defaults consistent with ms, which might be helpful for users.

Reply:
Done. Thanks for the input.